test/integration/idle_timeout_integration_test.cc

```diff
@@ -0,0 +1,161 @@
+#include "test/integration/http_protocol_integration.h"
+
+namespace Envoy {
+namespace {
+
+class IdleTimeoutIntegrationTest : public HttpProtocolIntegrationTest {
+public:
+  void initialize() override {
+    config_helper_.addConfigModifier(
+        [&](envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager& hcm)
+            -> void {
+          auto* route_config = hcm.mutable_route_config();
+          auto* virtual_host = route_config->mutable_virtual_hosts(0);
+          auto* route = virtual_host->mutable_routes(0)->mutable_route();
+          route->mutable_idle_timeout()->set_seconds(0);
+          route->mutable_idle_timeout()->set_nanos(TimeoutMs * 1000 * 1000);
+          // For validating encode100ContinueHeaders() timer kick.
+          hcm.set_proxy_100_continue(true);
+        });
+    HttpProtocolIntegrationTest::initialize();
+  }
+
+  IntegrationStreamDecoderPtr setupPerStreamIdleTimeoutTest() {
+    initialize();
+    codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+    auto encoder_decoder =
+        codec_client_->startRequest(Http::TestHeaderMapImpl{{":method", "GET"},
+                                                            {":path", "/test/long/url"},
+                                                            {":scheme", "http"},
+                                                            {":authority", "host"}});
+    request_encoder_ = &encoder_decoder.first;
+    auto response = std::move(encoder_decoder.second);
+    fake_upstream_connection_ = fake_upstreams_[0]->waitForHttpConnection(*dispatcher_);
+    upstream_request_ = fake_upstream_connection_->waitForNewStream(*dispatcher_);
+    upstream_request_->waitForHeadersComplete();
+    return response;
+  }
+
+  void Sleep() { std::this_thread::sleep_for(std::chrono::milliseconds(TimeoutMs / 2)); }
+
+  // TODO(htuch): This might require scaling for TSAN/ASAN/Valgrind/etc. Bump if
+  // this is the cause of flakes.
+  static constexpr uint64_t TimeoutMs = 200;
+};
+
+INSTANTIATE_TEST_CASE_P(Protocols, IdleTimeoutIntegrationTest,
+                        testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams()),
+                        HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+// Per-stream idle timeout when waiting for initial upstream headers.
+TEST_P(IdleTimeoutIntegrationTest, PerStreamIdleTimeoutBeforeUpstreamHeaders) {
```

Comment:
Can you audit these tests for flakes?

I think a bunch of them are subject to the failure mode I've been fixing where Envoy will reset both upstream and downstream because [timeout] and then there's a race in cleanUpstreamAndDownstream because if the FIN arrives when we post the close() it registers as an unexpected disconnect.
(Alternately I'd be happy to say we shouldn't detect unexpected disconnect in cleanup and fix it once and for all there, but I think @mattklein123's preference is to have things which are closed explicitly handled.)
I believe  fake_upstream_connection_->waitForDisconnect(); right after the code client closing will fix any of these.

Reply:
Will fix by adding `fake_upstreams_->set_allow_unexpected_disconnects(true);` to setup.